### PR TITLE
Transaction support

### PR DESCRIPTION
**Pull request by ilyasokay**
Transaction which is supported on Mongo v4 and above is almost ready. Just need 1-2 days to check completely. Thanks for your help @Smolevich .

### CODE REVIEW

**Comment by ilyasokay** [comment]
There is still work to do with handling deadlock in transaction, especially configuring exception messages of mongodb.

**Comment by Smolevich** [comment]
@ilyasokay do you have plans to continue work with PR?

**Comment by ilyasokay** [comment]
> @ilyasokay do you have plans to continue work with PR?

@Smolevich  Sorry for the delay, I had some urgent works to do; but I will continue this pr as soon as possible.

File: src/Jenssegers/Mongodb/Query/Builder.php [hunk 1]
```diff
@@ -833,7 +848,7 @@ protected function performUpdate($query, array $options = [])
         }
 
         $wheres = $this->compileWheres();
-        $result = $this->collection->UpdateMany($wheres, $query, $options);
+        $result = $this->collection->UpdateMany($wheres, $query, array_merge($options, $this->session()));
```
Comment (alcaeus) [inline src/Jenssegers/Mongodb/Query/Builder.php:851]: As per signature below:
```suggestion
        $result = $this->collection->UpdateMany($wheres, $query, $this->session($options));
```

There are a few more instances of this in the class.

File: src/Jenssegers/Mongodb/Query/Builder.php [hunk 2]
```diff
@@ -1153,6 +1168,15 @@ public function options(array $options)
         return $this;
     }
 
+    protected function session(array $options = [])
+    {
+        if ($session = $this->connection->getSession()) {
```
Comment (alcaeus) [inline src/Jenssegers/Mongodb/Query/Builder.php:1173]: Suggestion to not override an existing session:

```suggestion
        if (!isset($options['session']) && $session = $this->connection->getSession()) {
```

File: src/Jenssegers/Mongodb/Connection.php [hunk 3]
```diff
@@ -267,6 +272,40 @@ protected function getDefaultSchemaGrammar()
         return new Schema\Grammar();
     }
 
+    public function beginTransaction(array $options = [])
```
Comment (alcaeus) [inline src/Jenssegers/Mongodb/Connection.php:275]: What is the expected behaviour if a transaction is already in progress?

Comment (divine) [reply src/Jenssegers/Mongodb/Connection.php:275]: Hi @alcaeus , there is another PR which looks promising. If you'd like and have a time, please take a look https://github.com/jenssegers/laravel-mongodb/pull/1904

Thanks!

**Comment by divine** [comment]
Closing in favor of https://github.com/jenssegers/laravel-mongodb/pull/1904

**Pull request closed without merging**